SIMSBIOHUB-145: Survey Locations

al-rosenthal opened the pull request:
## Links to Jira Tickets

[SIMSBIOHUB-145](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-145)

## Description of Changes

- Migration to add new survey locations table
- updated API/ Services to handle multiple 'locations' for a single survey
- updated broken tests 
- refactored old code
- Updated front end to pass an array of locations for a survey

## Testing Notes

- You should be able to create/ edit surveys with a location same as you would before, boundaries and study area names should appear as expected
- The work done here is ground work for upcoming survey work

NickPhura committed:
Add migration to create new survey spatial tables and migrate existing data

NickPhura committed:
WIP: replace old survey spatial code

NickPhura committed:
update

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-145

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-145

NickPhura committed:
WIP: Update API for survey locations

NickPhura committed:
Updates

NickPhura committed:
FIx seed

NickPhura committed:
Fix seed

al-rosenthal committed:
added zod schema for survey location data

al-rosenthal committed:
updated form

al-rosenthal committed:
got create survey wired up to new tables

al-rosenthal committed:
wired up functions so things will load properly

al-rosenthal committed:
fixed post requests

al-rosenthal committed:
clean up

al-rosenthal committed:
fixing more map loads

al-rosenthal committed:
updated default spatial keyword to geojson

al-rosenthal committed:
Merge branch 'dev' into SIMSBIOHUB-145

al-rosenthal committed:
fixed app tests

al-rosenthal committed:
added update survey locations

al-rosenthal committed:
fixing api tests

al-rosenthal committed:
fixing more broken tests

al-rosenthal committed:
Merge branch 'dev' into SIMSBIOHUB-145

al-rosenthal committed:
cleaning up code smell

al-rosenthal committed:
Merge branch 'SIMSBIOHUB-145' of https://github.com/bcgov/biohubbc into SIMSBIOHUB-145

al-rosenthal committed:
cleaning up code, adding comments

al-rosenthal committed:
cleaned up console logs

al-rosenthal committed:
Merge branch 'dev' into SIMSBIOHUB-145

KjartanE reviewed (dismissed):
Tested locally, works as intended

NickPhura commented:
Getting some errors editing a survey.

Setup: I created a project and a survey, which all worked as expected.

---
Editing the survey location directly:

When I click save, the API throws a bunch of openapi validation errors for missing fields.

---
Editing the entire survey:

(This one seems unrelated to the location changes?)

When the edit survey page loads, it immediately pops a delete confirmation dialogue "Remove all strata?". 
If you cancel that dialogue, and then click the "Cancel" button to back out of the edit, the same dialogue pops up again.
![image](https://github.com/bcgov/biohubbc/assets/903781/49d600d8-916c-4351-b42a-bdd79c0ffae3)

If you cancel that dialogue, and then click save without making any changes, the survey end date field is throwing a yup validation error (I think we were seeing something like this in Macs PR at one point too, though I don't remember the issue/resolution now)
![image](https://github.com/bcgov/biohubbc/assets/903781/4ba47f7b-477a-4e45-b416-a9a12ec5e674)

---

I can help investigate these as well since I don't have anything else on the go.

Edit: I think I have all of these resolved locally, will push up once I check all of the tests are happy, etc.
Edit 2: Committed

NickPhura committed:
Updated site selection form to better handle the "delete stratum" warning dialogue when editing a survey.ts
Fixed a few openapi issues on the edit survey endpoints pertaining to the newer location section data.
Tweaked the date field components slightly to better handle validation on an empty field (gets set to null now, rather than 'invalid date')

NickPhura committed:
code smell

curtisupshall commented on api/src/repositories/site-selection-strategy-repository.ts:
```diff
@@ -53,7 +53,7 @@ export class SiteSelectionStrategyRepository extends BaseRepository {
       .select('ss.name')
       .from('survey_site_strategy as sss')
-      .where('sss.survey_id', 1)
+      .where('sss.survey_id', surveyId)
```
Good catch 😅

curtisupshall commented on api/src/paths/project/{projectId}/survey/{surveyId}/view.test.ts:
```diff
@@ -88,6 +95,7 @@ describe('survey/{surveyId}/view', () => {
 
         const response = responseValidator.validateResponse(200, apiResponse);
 
+        console.log(response);
```
Nit: console log

NickPhura replied:
Removed

curtisupshall commented on api/src/repositories/survey-location-repository.ts:
```diff
@@ -0,0 +1,96 @@
+  name: z.string(),
+  description: z.string(),
+  geography: z.any(),
+  geojson: z.any(),
```
Could we make this `z.object` (or `z.array`?) to be a little more specific? Are we going with `z.any` because this can be a JSON string _or_ JSON object?

curtisupshall replied:
I think it's beneficial to answer the question "is this a JSON string or is this a JSON object" as early on in the request as possible

NickPhura replied:
`geojson` is currently always an array of Feature, I believe. Though the column itself is just a generic jsonb type.

I've updated these definitions to be a bit more specific.

curtisupshall commented on api/src/services/survey-location-service.ts:
```diff
@@ -0,0 +1,33 @@
+import { SurveyLocationRecord, SurveyLocationRepository } from '../repositories/survey-location-repository';
+import { DBService } from './db-service';
+
+export class SurveyLocationService extends DBService {
```
Nit: jsdocs for class + methods

NickPhura replied:
Updated

curtisupshall commented on api/src/services/survey-service.ts:
```diff
@@ -98,9 +99,9 @@ export class SurveyService extends DBService {
-      site_selection: await this.siteSelectionStrategyService.getSiteSelectionDataBySurveyId(surveyId),
+      locations: await this.getSurveyLocationsData(surveyId),
       participants: await this.surveyParticipationService.getSurveyParticipants(surveyId),
+      site_selection: await this.siteSelectionStrategyService.getSiteSelectionDataBySurveyId(surveyId),
```
Good catch

curtisupshall commented on api/src/services/survey-service.ts:
```diff
@@ -485,6 +486,11 @@ export class SurveyService extends DBService {
     return surveyId;
   }
 
+  async insertSurveyLocations(surveyId: number, data: PostLocationData): Promise<void> {
```
Nit: jsdocs for new methods

NickPhura replied:
Added

curtisupshall commented on app/package-lock.json:
Nit: remove from diff:
```bash
git checkout dev -- app/package-lock.json
```

NickPhura replied:
Removed. Though it really doesn't matter too much, other than maybe being mildly annoying sometimes if it causes conflicts.

curtisupshall commented on app/src/components/fields/MultiAutocompleteField.tsx:
```diff
@@ -17,9 +22,15 @@ export interface IMultiAutocompleteField {
   required?: boolean;
   filterLimit?: number;
   chipVisible?: boolean;
+  onChange?: (
```
Very nice 😎

curtisupshall commented on app/src/components/fields/MultiAutocompleteField.tsx:
```diff
@@ -93,45 +103,46 @@ const MultiAutocompleteField: React.FC<IMultiAutocompleteField> = (props) => {
-      </Box>
-    );
+  const getExistingValue = (existingValues: (number | string)[]): IMultiAutocompleteFieldOption[] => {
+    if (existingValues) {
```
If `existingValues` is actually optional (hence this check), make the parameter optional:
```
const getExistingValue = (existingValues?: (number | string)[]): IMultiAutocompleteFieldOption[] => {
```
If `existingValues` isn't optional, remove this check (since it will always resolve `true`)

NickPhura replied:
Not sure the historical context for this check, but will add the ? for now, presuming the `if` was added for a good reason before.

curtisupshall commented on app/src/components/fields/MultiAutocompleteField.tsx:
```diff
@@ -93,45 +103,46 @@ const MultiAutocompleteField: React.FC<IMultiAutocompleteField> = (props) => {
+              checkedIcon={<CheckBox fontSize="small" />}
+              checked={selected}
+              // Always seem to be disabled
+              disabled={(props.options && props.options?.indexOf(renderOption) !== -1) || false}
```
Could probably use `props.options?.includes(renderOption)`

NickPhura replied:
For sure

Though tbh I am not really sure what this is even doing. Its trying to grey out options that aren't in the list of options, or something... Seems like an odd use case that we probably never actually encounter in our use cases. Probably just something copy/pasted from an example somewhere.

curtisupshall commented on app/src/components/fields/SingleDateField.tsx:
```diff
@@ -16,6 +18,10 @@ interface IDateProps {
   other?: TextFieldProps;
 }
 
+const CalendarIcon: React.FC = () => {
```
Nit: redundant

NickPhura replied:
Do you mean `React.FC`

curtisupshall replied:
I meant that `<Icon />` should just be used twice inline, rather than using `const CalendarIcon`

NickPhura replied:
Took it out anyways, as it wasn't doing anything

curtisupshall commented on app/src/features/surveys/components/StudyAreaForm.tsx:
```diff
@@ -5,25 +5,37 @@ import { useFormikContext } from 'formik';
+    {
+      survey_location_id: null as unknown as number,
+      name: '',
+      // TODO description is temporarily hardcoded until the new UI to populate this field is implemented
```
If there is an existing ticket to address this, it could be a good idea to link it

NickPhura replied:
Yup, I believe its this one in the current sprint..? https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-219

will add

curtisupshall reviewed (commented):
Code looks good! Lots of welcome bug fixes/improvements

NickPhura committed:
Code review comments

NickPhura commented:
@curtisupshall Pushed your review comment changes up (jumped the gun a bit and commented that they were updated before I actually committed the changes)

curtisupshall commented on app/src/components/fields/StartEndDateFields.tsx:
```diff
@@ -19,11 +19,11 @@ interface IStartEndDateFieldsProps {
 };
 
-const CalendarEndIcon: React.FC = () => {
+const CalendarEndIcon = () => {
```
Can we inline these Calendar icons?
```javascript
<DatePicker
  slots={{
    openPickerIcon: () => <Icon path={mdiCalendarEnd} size={1} />
  }}
```

curtisupshall commented on app/src/components/fields/SingleDateField.tsx:
```diff
@@ -40,6 +46,9 @@ const SingleDateField: React.FC<IDateProps> = (props) => {
     <LocalizationProvider dateAdapter={AdapterMoment}>
       <DatePicker
+        slots={{
+          openPickerIcon: CalendarIcon
```
Can we inline this Calendar icon?
```javascript
<DatePicker
  slots={{
    openPickerIcon: () => <Icon path={mdiCalendar} size={1} />
  }}
```

NickPhura replied:
Ill try, I feel like maybe there was some warning about inlining functions though, because react complains about it re-rendering every time (If I am remembering correctly).

curtisupshall replied:
Right, yeah I realized that after commenting. No worries if it can't be done

NickPhura replied:
Well, In-lined them, and it seems to be happy

NickPhura replied:
Wait spoke too soon, sonarcloud is complaining about it

curtisupshall replied:
Ahh too bad 😪

curtisupshall reviewed (dismissed):
Code looks good!

NickPhura committed:
Inline calendar icons

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-145

NickPhura committed:
Fix invalid date comparison

NickPhura committed:
revert calendar icons
